xdsl/dialects/csl/csl_wrapper.py

```diff
@@ -0,0 +1,308 @@
+
+    Additionally, any value yielded by the `layout_module` is passed as a block arg (lowered to `csl.param`)
+    to the program module. Order is important here as well. The layout module's `yield` op should be lowered to
+    `@set_tile_code`, while the program module's `yield` op should be discarded.
+
+    The layout module has two additional block args `x` and `y` as part of the `@set_tile_code` loop nest.
+    Operations using these args need to be lowered to the correct place in the loop nest.
+    """
+
+    name = "csl_wrapper.module"
+
+    width = prop_def(IntegerAttr)
+    height = prop_def(IntegerAttr)
+    params = prop_def(DictionaryAttr)
+
+    layout_module = region_def("single_block")
+    program_module = region_def("single_block")
+
+    @staticmethod
+    def from_properties(properties: dict[str, Attribute]):
+        layout_module = Region(Block())
+        program_module = Region(Block())
+        x = layout_module.block.insert_arg(IntegerType(16), 0)
+        y = layout_module.block.insert_arg(IntegerType(16), 1)
+        x.name_hint = "x"
+        y.name_hint = "y"
+
+        for name, value in properties.items():
+            if not isa(value, IntegerAttr[IntegerType]):
+                raise ValueError("Can only create module from IntegerAttr properties")
+            l_arg = layout_module.block.insert_arg(
+                value.type, len(layout_module.block.args)
+            )
+            l_arg.name_hint = name
+            p_arg = program_module.block.insert_arg(
+                value.type, len(program_module.block.args)
+            )
+            p_arg.name_hint = name
+
+        props = {
+            "width": properties.pop("width"),
+            "height": properties.pop("height"),
+            "params": DictionaryAttr(properties),
+        }
+
+        return ModuleOp(properties=props, regions=[layout_module, program_module])
+
+    def update_program_block_args_from_layout(self):
+        """Update `program_module` BlockArguments by adding yield op fields"""
+        assert (
+            len(self.program_module.block.args)
+            == len(self.layout_module.block.args) - 2
+            # minus two as layout_module has additional x and y args
+        ), "program_module block args should only contain args from properties when calling this function"
+
+        yield_op = self.layout_module.block.last_op
+        if not isinstance(yield_op, YieldOp):
+            raise ValueError("layout module must be terminated by csl_wrapper.yield")
+
+        if yield_op.fields is None:
+            raise ValueError("layout module yield must specify fields property")
+        yield_op.verify_()
+        for name, op in zip(yield_op.fields, yield_op.arguments):
+            arg = self.program_module.block.insert_arg(
+                op.type, len(self.program_module.block.args)
+            )
+            arg.name_hint = name.data
+
+    def verify_(self):
+        if "x" in self.properties or "y" in self.properties:
```

Comment:
The `x` and `y` params are reserved for loop variables of the `@set_tile_code` loop the layout_module needs to generate. Both are auto-generated and only appear as block args on the layout_module.

Reply:
I understand. What I'm saying is that I don't see a technical reason why we can't *also* have a parameter named `x` or `y` (or `height` or `width`). Disallowing certain names as parameters feels unintuitive, especially because there is no technical reason why we should.

Reply:
I'd like to refer to params by name in code generation/translation, which is what `get_program_arg()` and `get_layout_arg()` are for (I acty have code using these, though not in a PR yet). Having two instances of e.g. `x` would require a way to distinguish which is which and might be error prone. Reserving names seems cleanest and easiest.

Reply:
Why do we need `get_program_arg` *and* block arguments? I feel that one would suffice.